wikidata_game/util/database.php

```diff
@@ -0,0 +1,80 @@
+<?php declare(strict_types=1);
+
+require_once('helpers.php');
+
+/**
+ * Initiates and returns a database connection
+ *
+ * @param array $db_cnf List of required configuration to create the connection
+ * @return PDO The PDO database connection
+ * 
+ * @throws InvalidArgumentException when an invalid database configuration is passed
+ */
+// TODO: Unit Tests needed (Mocking Database connection)
+function getDb(array $db_cnf): PDO {
+    $required_keys = ['user', 'password', 'dbhost', 'dbname'];
+    
+    if(!array_has_keys($required_keys, $db_cnf)){
+        throw new InvalidArgumentException('Database configuration is missing one of the following required keys: ' 
+                                            . implode($required_keys, ', '));
+    }
+    
+    $dbhost = $db_cnf['dbhost'];
+    $dbname = $db_cnf['dbname'];
+    $dbuser = $db_cnf['user'];
+    $dbpass = $db_cnf['password'];
+
+    return new PDO('mysql:host=' . $dbhost . ';dbname=' . $dbname . ';charset=utf8', $dbuser, $dbpass);
+}
+
+/**
+ * Creates a closure to retrieve reference matches from a specified database connection
+ *
+ * @param PDO $db
+ * @return Closure function to query and retrieve matches
+ */
+// TODO: Unit Tests needed (Mocking Database connection)
+function createMatchesReader(PDO $db): Closure {
```

Comment:
When I first saw these I thought the functionality looked very similar to instantiating traditional classes in php (taking the parameter of the closure and using it in an inner 'method' looks very much like a constructor).

What would you think about actually making this more class like? An additional advantage of this is that traditional classes (and the layout of one class per file) is very well supported by the PHP ecosystem. I sketched out a little bit what this might possibly look like because I suspected that this ramble might not be that clear.

I stuck a branch (+PR) up here: https://github.com/wmde/reference-island/pull/65

This means that we can use the composer class auto-loader which avoids us creating these functions in the global namespace and allows us to better encapsulate them.

Reply:
You are of course completely correct, this is the functional module equivalent of a "method". I chose to write it this way, as it seemed less cumbersome to me to adopt at this moment than  a full Object Oriented approach. That being sade, I really like PR #65 . But I would prefder us to be consistent in that if we choose to work with classes, all of the patterns here should be converted to classes (And not just this specific one, with probably the exception of the small helper function).

Reply:
Yep, I suspect if we wanna go with classes then we should do all of them. The PR was very incomplete and just an attempt to sketch out what I meant :)

I'm generally leaning towards using classes and the autoloader as it seems a "sort-of PHP best practice" and I'm generally rather hesitant about having all the functions (and variables) effectively as globals.